File: javascript/triangle/triangle.js
export class Triangle {
  constructor(...sides) {
    this.sides = sides.sort((a, b) => a - b);
  }

  kind() {
    let uniq = new Set(this.sides);

  	if (!this.is_triangle()) throw new Error("Not a triangle");
  	if (uniq.size === 1) return 'equilateral';
	if (uniq.size === 2) return 'isosceles';
 	if (uniq.size === 3) return 'scalene';
  }

  is_triangle() {
  	const [a, b, c] = this.sides;
	if (a <= 0) return false;
	if (a + b < c) return false;

	return true;
  }
}

File: ruby/triangle/triangle.rb
class Triangle
  attr_reader :sides

  def initialize(sides)
    @sides = sides.sort
  end

  def equilateral?
    return false if !triangle?
    sides.uniq.size == 1
  end

  def isosceles?
    return false if !triangle?
    sides.uniq.size <= 2
  end

  def scalene?
    return false if !triangle?
    sides.uniq.size == 3
  end

  def triangle?
    sides.all? { |side| side > 0 } and sides[0] + sides[1] >= sides[2]
  end
end
